### PR TITLE
Add in importance p-values calculation for classification and regression

### PR DESCRIPTION
**Pull request by kmacdon**
Adds in the Janitza calculation for importance p-values from the  R package [ranger](https://rdrr.io/cran/ranger/src/R/importance.R) for the RangerForestClassifier class. 

I also attempted to add a version for the Regressor, but the method requires that some variables have negative importance values and none of the sklearn datasets had variables unimportant enough to have negative values. Happy to make changes if you have any additional suggestions/requirements.

### CODE REVIEW

File: skranger/ensemble/ranger_forest_classifier.py [hunk 1]
```diff
@@ -295,3 +297,28 @@ def predict_log_proba(self, X):
         """
         proba = self.predict_proba(X)
         return np.log(proba)
+
+    def get_importance_pvalues(self):
+        """Return p-values for variable importance using the fast method from Janitza et al. (2016)
+        """
+
+        if self.importance != "impurity_corrected":
+            raise Exception("P-values can only be calculated for impurity corrected importance values")
```
Comment (crflynn) [inline skranger/ensemble/ranger_forest_classifier.py:306]: Let's raise a ValueError here, indicating that one of the initialization parameters is invalid for this method.

Comment (crflynn) [reply skranger/ensemble/ranger_forest_classifier.py:306]: We can be more explicit with the message here saying:
```
p-values can only be calculated with importance parameter set to `impurity_corrected`
```
I'm looking at this code for reference: https://github.com/imbs-hl/ranger/blob/e8b05f47892bb4968c4e6057f68b35bcd0b52225/R/importance.R

File: skranger/ensemble/ranger_forest_classifier.py [hunk 2]
```diff
@@ -295,3 +297,28 @@ def predict_log_proba(self, X):
...
+
+        vimp = np.array(self.ranger_forest_["variable_importance"])
+        m1 = vimp[vimp < 0]
+        m2 = vimp[vimp == 0]
+
+        if len(m1) < 1:
+            raise Exception("No negative importance values found, cannot calculate p-values.")
```
Comment (crflynn) [inline skranger/ensemble/ranger_forest_classifier.py:313]:
```suggestion
        if len(m1) == 0:
            raise ValueError("No negative importance values found, cannot calculate p-values.")
```

File: skranger/ensemble/ranger_forest_classifier.py [hunk 3]
```diff
@@ -295,3 +297,28 @@ def predict_log_proba(self, X):
...
+        if len(m2) < 1:
+            vimp_dist = np.concatenate((m1, -m2))
```
Comment (crflynn) [inline skranger/ensemble/ranger_forest_classifier.py:315]: in this case is vimp_dist just equal to m1?

File: tests/ensemble/test_ranger_forest_classifier.py [hunk 4]
```diff
@@ -114,6 +114,23 @@ def test_importance(self, iris_X, iris_y, importance, scale_permutation_importan
...
+
+        if importance not in ["none", "impurity", "impurity_corrected", "permutation"]:
+            with pytest.raises(ValueError):
+                rfc.fit(digits_X, digits_y)
+            return
+
+        if not importance == "impurity_corrected":
+            with pytest.raises(Exception):
+                rfc.fit(digits_X, digits_y)
+                rfc.get_importance_pvalues()
```
Comment (crflynn) [inline tests/ensemble/test_ranger_forest_classifier.py:128]: to be explicit about where this exception is raised
```suggestion
            rfc.fit(digits_X, digits_y)
            with pytest.raises(ValueError):
                rfc.get_importance_pvalues()
```

**Review by crflynn — CHANGES_REQUESTED** [review]
A few notes here. The implementation looks a bit different from the ranger R code. Is it doing the same thing as here? https://github.com/imbs-hl/ranger/blob/e8b05f47892bb4968c4e6057f68b35bcd0b52225/R/importance.R#L124

File: skranger/ensemble/ranger_forest_classifier.py [hunk 5]
```diff
@@ -295,3 +297,28 @@ def predict_log_proba(self, X):
         """
         proba = self.predict_proba(X)
         return np.log(proba)
+
+    def get_importance_pvalues(self):
+        """Return p-values for variable importance using the fast method from Janitza et al. (2016)
+        """
+
```
Comment (crflynn) [inline skranger/ensemble/ranger_forest_classifier.py:304]: Since this requires training, let's add this check to the beginning of the method: https://scikit-learn.org/stable/modules/generated/sklearn.utils.validation.check_is_fitted.html

File: skranger/ensemble/ranger_forest_classifier.py (hunk 3 above)
Comment (kmacdon) [reply skranger/ensemble/ranger_forest_classifier.py:315]: That should be -m1 not -m2, since you're right it would just be m1 in that case.

**Comment by kmacdon** [comment]
> A few notes here. The implementation looks a bit different from the ranger R code. Is it doing the same thing as here? https://github.com/imbs-hl/ranger/blob/e8b05f47892bb4968c4e6057f68b35bcd0b52225/R/importance.R#L124

Yes,

**Comment by crflynn** [comment]
It seems that even with the same seed there are different ranger results on mac vs linux, such that for mac the results end up failing the test here. That seems like another issue that I need to take a look at.

What we can do for this test, however, is rather than using the digits dataset we can stick with iris and add a feature of random values to force negative (or zero) importance values. Locally, I pulled your branch and what I did was switch back to the iris dataset for the test and append a random feature column:

```python
    def test_importance_pvalues(self, iris_X, iris_y, importance):
        rfc = RangerForestClassifier(importance=importance)
        np.random.seed(42)
        iris_X = np.concatenate((iris_X, np.random.uniform(size=(iris_X.shape[0], 1))), 1)
        ...
```

with this I'm able to generate a negative importance value for that random feature and get the test to pass on either platform.

Additionally, we can append a constant feature to force a 0 importance.

We could create another fixture, `iris_X_mod`, used in place of `iris_X` for this test, which is modified to hit each of the cases here to get full coverage on the test.

```python
        if len(m1) == 0:  # no non-negative importances (feature set of all zeros)
            raise ValueError("No negative importance values found, cannot calculate p-values.")
        if len(m2) < 1:  # at least one negative and no 0 importances (iris_X with an extra random feature)
            vimp_dist = np.concatenate((m1, -m1))
        else:  # at least one negative and at least one 0 importance (iris_X with an extra random and extra constant feature)
            vimp_dist = np.concatenate((m1, -m1, m2))
```

This should give us full coverage in testing as well as fix the CI failures.

**Comment by crflynn** [comment]
Do you intend on making these changes? I can take this the rest of the way if you don't mind.

**Comment by kmacdon** [comment]
> Do you intend on making these changes? I can take this the rest of the way if you don't mind.

I just added in the text fixes you suggested. I'm not as familiar with pytest so let me know if there's any need to clean them up, but if they're good then I can add in the regressor function as well

**Comment by crflynn** [comment]
That looks good to me. You can do something similar for the regressor.

**Review by crflynn — APPROVED** [review]
Got us caught up here and everything looks good now. Thanks for adding this 💯
